### ISSUE TITLE
[API Change Proposal] Inappropriate API of MissingValueImputation

### ISSUE THREAD

**Issue by inejc** [haifengl/smile#380]
I started digging into the code and realized that missing value imputers don't have a learning stage exposed through the API. E.g. there is no `train` method, only the `impute` method, which calculates the required state and immediately transforms the input data. The consequence is that implementations of the interface can't be used in cross validation where we want to fit instances on training data and use them on both training and validation data. In addition to that we must ensure that no info from the validation set is used during training, e.g. calculating the mean value of a column from the whole dataset (training + validation) and then using it to impute values in the validation set is in other words cheating. I'd also add a copy parameter to the constructor, similar to what feature transformers have.

Example usage:
```java
CrossValidation split = new CrossValidation(n, k);

for (int i = 0; i < k; i++) {
    double[][] xTrain = Math.slice(x, split.train[i]);
    double[][] xTest = Math.slice(x, split.test[i]);

    AverageImputation imputer = new AverageImputation(true);
    imputer.train(xTrain);
    xTrain = imputer.impute(xTrain);
    xTest = imputer.impute(xTest);
    ...
}
```

Please let me know what your opinion is regarding the above, thank you.

**Comment by haifengl** [comment]
Smile started with Java 6 long time ago. The cross validation API has many limitation today due to the strong type of Java. I will change the API to leverage lambda, which is more flexible to add a `train` method/interface.

**Comment by inejc** [comment]
Would you accept the PR that fixes the current implementations?

**Comment by haifengl** [comment]
if you look into the algorithm details, training is not a natural part of many missing value imputation algorithms. Cross validation is more meaning for supervised learning.
